packages/typescript-estree/src/typedefs.ts

```diff
@@ -0,0 +1,1380 @@
+  type: 'JSXClosingElement';
+  name: JSXTagNameExpression;
+}
+
+export interface JSXClosingFragment extends NodeBase {
+  type: 'JSXClosingFragment';
+}
+
+export interface JSXElement extends NodeBase {
+  type: 'JSXElement';
+  openingElement: JSXOpeningElement;
+  closingElement: JSXClosingElement | null;
+  children: JSXChild[];
+}
+
+export interface JSXEmptyExpression extends NodeBase {
+  type: 'JSXEmptyExpression';
+}
+
+export interface JSXExpressionContainer extends NodeBase {
+  type: 'JSXExpressionContainer';
+  expression: Expression | JSXEmptyExpression;
+}
+
+export interface JSXFragment extends NodeBase {
+  type: 'JSXFragment';
+  openingFragment: JSXOpeningFragment;
+  closingFragment: JSXClosingFragment;
+  children: JSXChild[];
+}
+
+export interface JSXIdentifier extends NodeBase {
+  type: 'JSXIdentifier';
+  name: string;
+}
+
+export interface JSXMemberExpression extends NodeBase {
+  type: 'JSXMemberExpression';
+  name: string;
+}
+
+export interface JSXOpeningElement extends NodeBase {
+  type: 'JSXOpeningElement';
+  typeParameters?: TSTypeParameterInstantiation;
+  selfClosing: boolean;
+  name: JSXTagNameExpression;
+  attributes: JSXAttribute[];
+}
+
+export interface JSXOpeningFragment extends NodeBase {
+  type: 'JSXOpeningFragment';
+}
+
+export interface JSXSpreadAttribute extends NodeBase {
+  type: 'JSXSpreadAttribute';
+  argument: Expression;
+}
+
+export interface JSXSpreadChild extends NodeBase {
+  type: 'JSXSpreadChild';
+  expression: Expression | JSXEmptyExpression;
+}
+
+export interface JSXText extends NodeBase {
+  type: 'JSXText';
+  value: string;
+  raw: string;
+}
+
+export interface LabeledStatement extends NodeBase {
+  type: 'LabeledStatement';
+  label: Identifier;
+  body: Statement;
+}
+
+export interface Literal extends LiteralBase {
+  type: 'Literal';
+}
+
+export interface LogicalExpression extends BinaryExpressionBase {
+  type: 'LogicalExpression';
+}
+
+export interface MemberExpression extends NodeBase {
+  type: 'MemberExpression';
+  object: LeftHandSideExpression;
+  property: Expression | Identifier;
+  computed?: boolean;
+}
+
+export interface MetaProperty extends NodeBase {
+  type: 'MetaProperty';
+  meta: Identifier;
+  property: Identifier;
+}
+
+export interface MethodDefinition extends MethodDefinitionBase {
+  type: 'MethodDefinition';
+}
+
+export interface NewExpression extends NodeBase {
+  type: 'NewExpression';
+  callee: LeftHandSideExpression;
+  arguments: Expression[];
+  typeParameters?: TSTypeParameterInstantiation;
+}
+
+export interface ObjectExpression extends NodeBase {
+  type: 'ObjectExpression';
+  properties: ObjectLiteralElementLike[];
+}
+
+export interface ObjectPattern extends NodeBase {
+  type: 'ObjectPattern';
+  properties: ObjectLiteralElementLike[];
+  typeAnnotation?: TSTypeAnnotation;
+  optional?: boolean;
+}
+
+export interface Program extends NodeBase {
+  type: 'Program';
+  body: Statement[];
+  sourceType: 'module' | 'script';
```

Comment:
`Program` node should have
```
  tokens?: Token[];
  comments?: Comment[];
```

and there is no tokens node
```
export interface Token extends BaseNode {
  value: string;
  regex?: {
    pattern: string;
    flags: string;
  };
  object?: any;
  property?: any;
  name?: any;
}
```

Reply:
We never add those two in `convert`, I assume they're added by eslint?

Reply:
they are added after `convert` base on flags `comment` and `tokens`
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/ast-converter.ts#L44

Reply:
cool, I missed that!
Note that we never add `object`, `property`, or `name` to tokens.

Reply:
ahh it was there before i refactored some parts of code, right now its no longer present :>

Reply:
i found out why i added it there:
see https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/typescript-estree/src/convert.ts#L338